fe/fe-core/src/main/java/com/starrocks/planner/PlanNode.java

```diff
@@ -710,16 +712,132 @@ public boolean canPushDownRuntimeFilter() {
     public void checkRuntimeFilterOnNullValue(RuntimeFilterDescription description, Expr probeExpr) {
     }
 
-    public boolean pushDownRuntimeFilters(RuntimeFilterDescription description, Expr probeExpr) {
+    /**
+     * Return candidate slot exprs which is same to the expr, eg: tb1 join tb2 on tb1.a = tb2.b, when expr param is tb1.a,
+     * tb2.b is the candidate slot expr for tb1.a which has the same syntax for the query.
+     *
+     * @param expr: the slot expr that need to find its candidate slot exprs.
+     * @return List<Expr>: all the slot expr's candidate slot exprs.
+     */
+    public List<Expr> candidatesOfSlotExpr(Expr expr) {
+        List<Expr> candidateSlotExprs = Lists.newArrayList();
+        // NOTE: No need to check expr is slot or not here, each node should implement its `candidatesOfSlotExpr` itself.
+        if (!expr.isBoundByTupleIds(getTupleIds())) {
+            return candidateSlotExprs;
+        }
+        // only contain the input expr by default.
+        candidateSlotExprs.add(expr);
+        return candidateSlotExprs;
+    }
+
+    public List<List<Expr>> candidatesOfSlotExprs(List<Expr> exprs) {
+        List<List<Expr>> candidatesOfSlotExprs = Lists.newArrayList();
+        for (Expr expr: exprs) {
+            List<Expr> candidates = candidatesOfSlotExpr(expr);
+            if (candidates.isEmpty()) {
+                return Lists.newArrayList();
+            }
+            candidatesOfSlotExprs.add(candidates);
+        }
+        return permutaionsOfPartitionByExprs(candidatesOfSlotExprs);
+    }
+
+    public static List<List<Expr>> permutaionsOfPartitionByExprs(List<List<Expr>> partitionByExprs) {
+        if (partitionByExprs.isEmpty()) {
+            return Lists.newArrayList();
+        }
+
+        int totalCount = 1;
+        for (List<Expr> exprs: partitionByExprs) {
+            totalCount *= exprs.size();
+        }
+
+        List<List<Expr>> permutations = Lists.newArrayList();
+        boolean isGreaterThanExpect = totalCount > 16;
+        if (isGreaterThanExpect) {
+            totalCount = 16;
+        }
+
+        if (totalCount <= 0) {
+            return Lists.newArrayList();
+        }
+
+        for (int i = 0; i < totalCount; i++) {
+            permutations.add(Lists.newArrayList());
+        }
+
+        for (int k = 0; k < partitionByExprs.size() - 1; k++) {
+            List<Expr> exprs = partitionByExprs.get(k);
+            int numTimes = totalCount / exprs.size();
+            for (int i = 0; i < exprs.size(); i++) {
+                for (int j = 0; j < numTimes; j++) {
+                    if ((i * numTimes) + j < totalCount) {
+                        permutations.get((i * numTimes) + j).add(exprs.get(i));
+                    }
+                }
+            }
+        }
+        // output the last column
+        List<Expr> exprs = partitionByExprs.get(partitionByExprs.size() - 1);
```

Comment:
this permutation algorithm are vary hard to read.
partition_by_exprs' size can to be too big, it is very be 2, 3, 4, 5; if it is large than 5, we can discard to tackle with this situtations.
so when we can use a table-driven method to permuation a list.
size=2 => [0,1], [1, 0]
size=3 => [0,1,2], [0,2,1],[2,0,1], [1,0,2], [1,2,0], [2,1,0]

writing-down a literal permutation of list indices looks so stupid, however we can write a recurisve function to generate   permutation of indices.

for an example:

perm(0)= listOf(0);
perm(m) {
   ans=perm(m-1);
   return range(1..n).map{i->ans.subList(0, i) + [m] + ans.subList(i, m-1);
}

Reply:
done.